magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,678 @@
+  """
+  def __init__(self, state, xml_score_part=None):
+    self.xml_score_part = xml_score_part
+    self.part_name = ""
+    self.midi_channel = 1
+    self.midi_program = 1
+    self.state = state
+    if xml_score_part != None:
+      self.parse()
+
+  def parse(self):
+    """Parse the <score-part> element to an in-memory representation"""
+    if self.xml_score_part.find("part-name") != None:
+      self.part_name = self.xml_score_part.find("part-name").text
+
+    xml_midi_instrument = self.xml_score_part.find("midi-instrument")
+    if xml_midi_instrument != None and \
+      xml_midi_instrument.find("midi-channel") != None and \
+      xml_midi_instrument.find("midi-program") != None:
+      self.midi_channel = int(xml_midi_instrument.find("midi-channel").text)
+      self.midi_program = int(xml_midi_instrument.find("midi-program").text)
+    else:
+      # If no MIDI info, increment MIDI channel and use default program
+      self.midi_channel = self.state.midi_channel + 1
+      self.state.midi_channel = self.midi_channel
+      self.midi_program = 1
+
+  def __str__(self):
+    score_str = "ScorePart: " + self.part_name
+    score_str += ", Channel: " + str(self.midi_channel)
+    score_str += ", Program: " + str(self.midi_program)
+    return score_str
+
+
+class Part(object):
+  """Internal represention of a MusicXML <part> element"""
+  def __init__(self, xml_part, score_part, state):
+    self.xml_part = xml_part
+    self.score_part = score_part
+    self.measures = []
+    self.transposes = False
+    self.state = state
+    self.parse()
+
+  def parse(self):
+    """Parse the <part> element"""
+
+    # Reset the time position when parsing each part
+    #CURRENT_TIME_POSITION = 0
+    self.state.time_position = 0
+    self.state.midi_channel = self.score_part.midi_channel
+    self.state.midi_program = self.score_part.midi_program
+    self.state.transpose = 0
+
+    xml_measures = self.xml_part.findall("measure")
+    for child in xml_measures:
+      measure = Measure(child, self, self.state)
+      self.measures.append(measure)
+
+  def getnotesinvoice(self, voice):
+    """Get all the notes in this part for the specified voice number"""
+    notes = []
+    for measure in self.measures:
+      for note in measure.notes:
+        if note.voice == voice:
+          notes.append(note)
+
+    return notes
+
+  def __str__(self):
+    part_str = "Part: " + self.score_part.part_name
+    return part_str
+
+
+class Measure(object):
+  """Internal represention of the MusicXML <measure> element"""
+  def __init__(self, xml_measure, part, state):
+    self.xml_measure = xml_measure
+    self.notes = []
+    self.tempos = []
+    self.time_signature = None
+    self.key_signature = KeySignature(state) # Default to C major
+    self.current_ticks = 0      # Cumulative tick counter for this measure
+    self.transpose = 0          # Number of semitones to transpose notes
```

Comment:
I'm honestly not sure. I put this property here because the <measure> element is where the transpose property is located in MusicXML. Whether this needs to be part of the Measure class or not is contingent upon the question you asked above.